libcloud/compute/drivers/packet.py

```diff
@@ -0,0 +1,254 @@
+        return list(map(self._to_size, data))
+
+    def create_node(self, project_id, name, size, image, location):
+        """
+        Create a node.
+
+        :return: The newly created node.
+        :rtype: :class:`Node`
+        """
+
+        params = {'hostname': name, 'plan': size.id,
+                  'operating_system': image.id, 'facility': location.id,
+                  'include': 'plan', 'billing_cycle': 'hourly'}
+
+        data = self.connection.request('/projects/%s/devices' % (project_id),
+                                       params=params, method='POST')
+
+        status = data.object.get('status', 'OK')
+        if status == 'ERROR':
+            message = data.object.get('message', None)
+            error_message = data.object.get('error_message', message)
+            raise ValueError('Failed to create node: %s' % (error_message))
+        return self._to_node(data=data.object)
+
+    def reboot_node(self, node):
+        params = {'type': 'reboot'}
+        res = self.connection.request('/devices/%s/actions' % (node.id),
+                                      params=params, method='POST')
+        return res.status == httplib.OK
+
+    def destroy_node(self, node):
+        res = self.connection.request('/devices/%s' % (node.id),
+                                      method='DELETE')
+        return res.status == httplib.OK
+
+    def list_ssh_keys(self):
+        """
+        List all the available SSH keys.
+
+        :return: Available SSH keys.
+        :rtype: ``list`` of :class:`SSHKey`
+        """
+        data = self.connection.request('/ssh-keys').object['ssh_keys']
+        return list(map(self._to_ssh_key, data))
+
+    def create_ssh_key(self, name, ssh_key):
+        """
+        Create a new SSH key.
+
+        :param      name: Key name (required)
+        :type       name: ``str``
+
+        :param      name: Valid public key string (required)
+        :type       name: ``str``
+        """
+        params = {'label': name, 'key': ssh_key}
+        data = self.connection.request('/ssh-keys', method='POST',
+                                       params=params).object
+        return self._to_ssh_key(data)
+
+    def destroy_ssh_key(self, key_id):
+        """
+        Delete an existing SSH key.
+
+        :param      key_id: SSH key id (required)
+        :type       key_id: ``str``
+        """
+        res = self.connection.request('/ssh-keys/%s' % (key_id),
+                                      method='DELETE')
+        return res.status == httplib.NO_CONTENT
+
+    def _to_node(self, data):
+        extra_keys = ['created_at', 'updated_at',
+                      'userdata', 'billing_cycle', 'locked']
+        if 'state' in data:
+            state = self.NODE_STATE_MAP.get(data['state'], NodeState.UNKNOWN)
+        else:
+            state = NodeState.UNKNOWN
+
+        public_ips = []
+        private_ips = []
+        if 'ip_addresses' in data and data['ip_addresses'] is not None:
```

Comment:
This could be made a bit more readable if you refactor it in a new method - e.g. `_get_ips_for_node` or similar.